sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -651,6 +651,10 @@ object ColumnPruning extends Rule[LogicalPlan] {
     // Can't prune the columns on LeafNode
     case p @ Project(_, _: LeafNode) => p
 
+    // Vectorized gapply in R should not prune columns because all referred fields
+    // can be used within given R native function.
+    case p @ Project(_, _: FlatMapGroupsInRWithArrow) => p
```

Comment:
Before this, is `FlatMapGroupsInRWithArrow` be considered to use only grouping column? For example, in the example of PR description, the column `gear`?

Reply:
There might be some more grouping columns. For example, 
```r
df <- createDataFrame(mtcars)
explain(count(groupBy(gapply(df,
                             c("gear", "hp"),
                             function(key, group) {
                               data.frame(gear = key[[1]], disp = mean(group$disp))
                             },
                             structType("gear double, disp double")))), TRUE)
```

Before:

```
== Optimized Logical Plan ==
Aggregate [count(1) AS count#42L]
+- Project
   +- FlatMapGroupsInRWithArrow [StructField(mpg,DoubleType,true), StructField(cyl,DoubleType,true), StructField(disp,DoubleType,true), StructField(hp,DoubleType,true), StructField(drat,DoubleType,true), StructField(wt,DoubleType,true), StructField(qsec,DoubleType,true), StructField(vs,DoubleType,true), StructField(am,DoubleType,true), StructField(gear,DoubleType,true), StructField(carb,DoubleType,true)], [StructField(gear,DoubleType,true), StructField(disp,DoubleType,true)], createexternalrow(gear#9, hp#3, StructField(gear,DoubleType,true), StructField(hp,DoubleType,true)), [gear#9, hp#3]
      +- Project [hp#3, gear#9]
         +- LogicalRDD [mpg#0, cyl#1, disp#2, hp#3, drat#4, wt#5, qsec#6, vs#7, am#8, gear#9, carb#10], false

== Physical Plan ==
*(4) HashAggregate(keys=[], functions=[count(1)], output=[count#42L])
+- Exchange SinglePartition
   +- *(3) HashAggregate(keys=[], functions=[partial_count(1)], output=[count#45L])
      +- *(3) Project
         +- FlatMapGroupsInRWithArrow [...], [StructField(mpg,DoubleType,true), StructField(cyl,DoubleType,true), StructField(disp,DoubleType,true), StructField(hp,DoubleType,true), StructField(drat,DoubleType,true), StructField(wt,DoubleType,true), StructField(qsec,DoubleType,true), StructField(vs,DoubleType,true), StructField(am,DoubleType,true), StructField(gear,DoubleType,true), StructField(carb,DoubleType,true)], [gear#35, disp#36], createexternalrow(gear#9, hp#3, StructField(gear,DoubleType,true), StructField(hp,DoubleType,true)), [gear#9, hp#3]
            +- *(2) Sort [gear#9 ASC NULLS FIRST, hp#3 ASC NULLS FIRST], false, 0
               +- Exchange hashpartitioning(gear#9, hp#3, 200)
                  +- *(1) Project [hp#3, gear#9]
                     +- *(1) Scan ExistingRDD arrow[mpg#0,cyl#1,disp#2,hp#3,drat#4,wt#5,qsec#6,vs#7,am#8,gear#9,carb#10]
```

After:

```
== Optimized Logical Plan ==
Aggregate [count(1) AS count#42L]
+- Project
   +- FlatMapGroupsInRWithArrow [StructField(mpg,DoubleType,true), StructField(cyl,DoubleType,true), StructField(disp,DoubleType,true), StructField(hp,DoubleType,true), StructField(drat,DoubleType,true), StructField(wt,DoubleType,true), StructField(qsec,DoubleType,true), StructField(vs,DoubleType,true), StructField(am,DoubleType,true), StructField(gear,DoubleType,true), StructField(carb,DoubleType,true)], [StructField(gear,DoubleType,true), StructField(disp,DoubleType,true)], createexternalrow(gear#9, hp#3, StructField(gear,DoubleType,true), StructField(hp,DoubleType,true)), [gear#9, hp#3]
      +- LogicalRDD [mpg#0, cyl#1, disp#2, hp#3, drat#4, wt#5, qsec#6, vs#7, am#8, gear#9, carb#10], false

== Physical Plan ==
*(4) HashAggregate(keys=[], functions=[count(1)], output=[count#42L])
+- Exchange SinglePartition
   +- *(3) HashAggregate(keys=[], functions=[partial_count(1)], output=[count#45L])
      +- *(3) Project
         +- FlatMapGroupsInRWithArrow [...], [StructField(mpg,DoubleType,true), StructField(cyl,DoubleType,true), StructField(disp,DoubleType,true), StructField(hp,DoubleType,true), StructField(drat,DoubleType,true), StructField(wt,DoubleType,true), StructField(qsec,DoubleType,true), StructField(vs,DoubleType,true), StructField(am,DoubleType,true), StructField(gear,DoubleType,true), StructField(carb,DoubleType,true)], [gear#35, disp#36], createexternalrow(gear#9, hp#3, StructField(gear,DoubleType,true), StructField(hp,DoubleType,true)), [gear#9, hp#3]
            +- *(2) Sort [gear#9 ASC NULLS FIRST, hp#3 ASC NULLS FIRST], false, 0
               +- Exchange hashpartitioning(gear#9, hp#3, 200)
                  +- *(1) Scan ExistingRDD arrow[mpg#0,cyl#1,disp#2,hp#3,drat#4,wt#5,qsec#6,vs#7,am#8,gear#9,carb#10]
```

and .. this crestes new output attributes from the given output schema (this is similar with Pandas groupped map UDF).

Reply:
If `FlatMapGroupsInRWithArrow` refers all outputs from its child plan, I think we don't need to add special rule for it. Can we avoid pruning by doing this?

Reply:
I think the plan shows a bit confusing about its output. I followed existing `FlatMapGroupsInRExec` to minimize review overhead and diff but .. I might have to fix it.

Reply:
But output attributes should be new attributes and I wasn't sure about associating child's output to the current plans output since they cannot be chained by expressions.

Reply:
Oh, I see. `FlatMapGroupsInRWithArrow ` has only one expression `keyDeserializer` which refers to grouping columns. For other columns in the group, it is missed from references of this plan. So this fix looks good.